src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -143,25 +143,24 @@ QSizeF QgsLayerTreeModelLegendNode::drawSymbol( const QgsLegendSettings &setting
 
   if ( ctx && ctx->painter )
   {
+    const QgsScopedRenderContextScaleToPixels scopedScaleToPixels( *( ctx->context ) );
+    const double scaleFactor = ctx->context->scaleFactor();
```

Comment:
Why do we need to multiply by the scale factor here? Won't this affect other symbols too?

Reply:
Up to this point the nodes of this type have only been assigned a simple QIcon (square solid color rectangle). Under those circumstances it does not really matter if a scale factor is applied. (A square solid color rectangle more or less remains a square solid color rectangle regardless of scaling). But if a more complex QIcon is assigned (e.g. arrow) without scaling then there will be unwanted distortion/pixelation. If you look at the new /expected_legend_diagram_attributes/expected_legend_diagram_attributes_mask.png I uploaded, you will find that the icons are now slightly more rectangular because proper scaling is applied. I am guessing no one else ever bothered to do so because it was never really noticeable.

Reply:
Watch out, `ctx->context` may be nullptr here. You'll need to check that and handle gracefully.